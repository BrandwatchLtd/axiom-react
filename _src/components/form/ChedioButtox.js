import React, { Component } from 'react';
import classnames from 'classnames';
import { enhance, addPropTypes, addClassName } from '../_utils/components';

if (__INCLUDE_CSS__) {
  require('./ChedioButtox.scss');
}

export class ChedioButtox extends Component {
  static propTypes = {
    children: { node: true },
    inline: { bool: true },
    inputClassName: { string: true },
    inputType: { string: true, isRequired: true },
    labelClassName: { string: true },
  };

  render() {
    const {
      className,
      children,
      inline,
      inputClassName,
      inputType,
      labelClassName,
      ...rest,
    } = this.props;

    const classes = classnames(className,
      `ax-chedio-buttox__group`, {
        [`ax-chedio-buttox__group--inline`]: inline === true,
        [`ax-chedio-buttox__group--space`]: !!children,
      }
    );

    return (
      <label className={ classes }>
        <input className={ classnames('ax-chedio-buttox', inputClassName) } type={ inputType } { ...rest } />
        <span className={ classnames('ax-chedio-buttox__label', labelClassName) }>
          { children }
        </span>
      </label>
    );
  }
}

export default enhance(ChedioButtox)(
  addPropTypes('global'),
  addClassName('global'),
);
