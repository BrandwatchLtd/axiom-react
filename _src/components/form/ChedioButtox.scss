@import '../../design-patterns/animations/mixins';
@import '../../design-patterns/animations/_vars.js';
@import '../../design-patterns/spacing/mixins';
@import '../../design-patterns/spacing/_vars.js';
@import '../icon/mixins';
@import '../icon/vars';
@import '../form/_vars.js';
@import '../typography/_vars.js';


.ax-chedio-buttox__group {
  @include space-block();

  display: flex;
  vertical-align: middle;
}

.ax-chedio-buttox__group--inline {
  @include space-inline();

  display: inline-flex;
}

.ax-chedio-buttox__group--space {
  .ax-chedio-buttox__label {
    &::before {
      margin: 0 $component-margin-h-sm 0 0;
    }
  }
}

.ax-chedio-buttox {
  position: absolute;
  opacity: 0;
  z-index: -1;

  &:disabled {
    + .ax-chedio-buttox__label {
      cursor: default;
    }
  }

  + .ax-chedio-buttox__label {
    @include transition((background-color, border-color, color), $transition-time-fast);

    display: flex;
    align-items: flex-start;
    line-height: $form-checkbox-size;
    cursor: pointer;

    &::before {
      @include transition((background-color, border-color, color), $transition-time-fast);

      display: flex;
      flex: none;
      align-items: center;
      justify-content: center;
      width: $form-checkbox-size;
      height: $form-checkbox-size;
      border-width: $form-input-border-width;
      border-style: solid;
      font-size: $font-size-sm;
    }
  }
}

// Radio and Check box: Colouring
.ax-chedio-buttox {
  + .ax-chedio-buttox__label {
    &::before {
      border-color: $input-border-color;
      background-color: $input-background-color;
      color: transparent;
    }
  }

  &:hover {
    + .ax-chedio-buttox__label {
      &::before {
        color: $input-placeholder-color;
      }
    }
  }

  &:focus {
    + .ax-chedio-buttox__label {
      &::before {
        border-color: $input-focus-color;
      }
    }
  }

  &:checked {
    + .ax-chedio-buttox__label {
      &::before {
        color: $input-checked-color;
      }
    }
  }

  &:disabled {
    + .ax-chedio-buttox__label {
      cursor: default;

      &::before {
        background-color: $input-disabled-color;
        color: transparent;
      }
    }
  }
}
