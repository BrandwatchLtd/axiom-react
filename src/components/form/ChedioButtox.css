:root {
  --cmp-chedio-indicator-space: calc(var(--spacing-grid-size) * 2);
}

.ax-chedio-buttox__container {
  display: flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
}

.ax-chedio-buttox__container--disabled {
  cursor: default;
}

.ax-chedio-buttox {
  position: absolute;
  opacity: 0;
  z-index: -1;
}

.ax-chedio-buttox__indicator {
  position: relative;
  width: var(--cmp-input-checkbox-size);
  height: var(--cmp-input-checkbox-size);
  border: var(--component-border-width) solid var(--color-theme-border);
  background-color: var(--color-theme-background);
  transition-property: background-color, border-color, color;
  transition-duration: var(--transition-time-base);
  transition-timing-function: var(--transition-function);

  @nest .ax-chedio-buttox:focus + &    { border-color: var(--cmp-input-border-color-focused); }
  @nest .ax-chedio-buttox:disabled + & { background-color: var(--color-theme-background--disabled); }

  &::before {
    content: '';
  }
}

.ax-chedio-buttox__label {
  margin-left: var(--cmp-chedio-indicator-space);
}

.ax-chedio-buttox__container--invalid {
  & .ax-chedio-buttox + .ax-chedio-buttox__indicator,
  & .ax-chedio-buttox:focus + .ax-chedio-buttox__indicator {
    border-color: var(--cmp-input-border-color-invalid);
  }
}
