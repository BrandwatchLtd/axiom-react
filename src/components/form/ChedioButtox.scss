@import '../../design-patterns/animations/mixins';
@import '../../design-patterns/animations/_vars.js';
@import '../../design-patterns/spacing/mixins';
@import '../../design-patterns/spacing/_vars.js';
@import '../form/vars';
@import '../typography/vars';

$checkbox-size: $line-height-body-px + ($component-border-width * 2);

.ax-chedio-buttox__container {
  @include space-block();

  display: flex;
  vertical-align: middle;
}

.ax-chedio-buttox__container--space {
  .ax-chedio-buttox__label {
    &::before {
      margin: 0 $component-margin-sm 0 0;
    }
  }
}

.ax-chedio-buttox {
  position: absolute;
  opacity: 0;
  z-index: -1;

  &:disabled {
    + .ax-chedio-buttox__label {
      cursor: default;
    }
  }

  + .ax-chedio-buttox__label {
    @include transition((background-color, border-color, color), $transition-time-fast);

    display: flex;
    align-items: flex-start;
    line-height: $checkbox-size;
    cursor: pointer;

    &::before {
      @include transition((background-color, border-color, color), $transition-time-fast);

      display: flex;
      flex: none;
      align-items: center;
      justify-content: center;
      width: $checkbox-size;
      height: $checkbox-size;
      border-width: $component-border-width;
      border-style: solid;
      font-size: $font-size-sm;
    }
  }
}

// Radio and Check box: Colouring
.ax-chedio-buttox {
  + .ax-chedio-buttox__label {
    &::before {
      border-color: $input-border-color;
      background-color: $input-background-color;
      color: transparent;
    }
  }

  &:hover {
    + .ax-chedio-buttox__label {
      &::before {
        color: $input-placeholder-color;
      }
    }
  }

  &:focus {
    + .ax-chedio-buttox__label {
      &::before {
        border-color: $input-focus-color;
      }
    }
  }

  &:checked {
    + .ax-chedio-buttox__label {
      &::before {
        color: $input-checked-color;
      }
    }
  }

  &:disabled {
    + .ax-chedio-buttox__label {
      cursor: default;

      &::before {
        background-color: $input-disabled-color;
        color: transparent;
      }
    }
  }
}
