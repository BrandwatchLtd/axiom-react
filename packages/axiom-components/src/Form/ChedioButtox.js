import PropTypes from "prop-types";
import React, { Component } from "react";
import classnames from "classnames";
import Base from "../Base/Base";
import Tick from "./Tick";
import Indeterminate from "./Indeterminate";

import "./ChedioButtox.css";

export default function ChedioButtox({
  children,
  className,
  disabled,
  inputType,
  invalid,
  title,
  onClick,
  indeterminate,
  checked,
  ...rest
}) {
  const getMarker = () => {
    if (indeterminate) {
      return <Indeterminate />;
    }

    if (checked) {
      return <Tick />;
    }
  };

  const getAriaChecked = () => {
    if (indeterminate) {
      return "mixed";
    }

    if (checked) {
      return true;
    }

    return false;
  };

  const handleClick =
    onClick &&
    ((e) => {
      if (e.target.tagName !== "INPUT") {
        e.stopPropagation();
        return;
      }
      onClick(e);
    });

  const classes = classnames(className, {
    [`${className}--disabled`]: disabled,
    [`${className}--invalid`]: invalid,
    [`${className}--indeterminate`]: indeterminate,
  });

  return (
    <Base
      Component="label"
      className={classes}
      onClick={handleClick}
      space="x2"
      title={title}
    >
      <input
        {...rest}
        checked={checked}
        className={`${className}__input`}
        disabled={disabled}
        type={inputType}
        aria-checked={getAriaChecked()}
      />

      <span className={`${className}__indicator`}>
        {inputType === "checkbox" && getMarker()}
      </span>
      {children && <span className={`${className}__label`}>{children}</span>}
    </Base>
  );
}

ChedioButtox.propTypes = {
  checked: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  indeterminate: PropTypes.bool,
  inputType: PropTypes.oneOf(["checkbox", "radio"]).isRequired,
  invalid: PropTypes.bool,
  onClick: PropTypes.func,
  title: PropTypes.string,
};
