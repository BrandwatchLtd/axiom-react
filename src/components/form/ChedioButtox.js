import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import Base from '../base/Base';

if (__INCLUDE_CSS__) {
  require('./ChedioButtox.scss');
}

export default class ChedioButtox extends Component {
  static propTypes = {
    children: PropTypes.node,
    inputClassName: PropTypes.string,
    inputType: PropTypes.string.isRequired,
    labelClassName: PropTypes.string,
  };

  render() {
    const {
      className,
      children,
      inputClassName,
      inputType,
      labelClassName,
      ...rest
    } = this.props;

    const classes = classnames(className,
      'ax-chedio-buttox__container', {
        ['ax-chedio-buttox__container--space']: !!children,
      }
    );

    return (
      <Base Component="label" className={ classes }>
        <input { ...rest } className={ classnames('ax-chedio-buttox', inputClassName) } type={ inputType } />
        <span className={ classnames('ax-chedio-buttox__label', labelClassName) }>
          { children }
        </span>
      </Base>
    );
  }
}
