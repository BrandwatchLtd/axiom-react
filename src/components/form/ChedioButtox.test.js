import React from 'react';
import renderer from 'react-test-renderer';
import ChedioButtox from './ChedioButtox';

const defaultProps = {
  inputType: 'checkbox',
  indicatorClassName: 'ax-lorem',
  inputClassName: 'ax-ipsum',
};

function getComponent(props = {}) {
  return renderer.create(
    <ChedioButtox
        { ...defaultProps }
        { ...props }>
      Lorem ipsum
    </ChedioButtox>
  );
}

describe('ChedioButtox', () => {
  it('renders with defaultProps', () => {
    const component = getComponent();
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders when disabled', () => {
    const component = getComponent({ disabled: true });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
