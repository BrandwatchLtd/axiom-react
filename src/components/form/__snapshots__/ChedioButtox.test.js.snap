// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChedioButtox renders with defaultProps 1`] = `
<label
  className="ax-chedio-buttox__container ax-space--x2"
  onClick={undefined}
  title="Lorem ipsum dolor sit amet"
>
  <input
    className="ax-chedio-buttox ax-ipsum"
    disabled={undefined}
    type="checkbox"
  />
  <span
    className="ax-chedio-buttox__indicator ax-lorem"
  />
  <span
    className="ax-chedio-buttox__label"
  >
    Lorem ipsum
  </span>
</label>
`;

exports[`ChedioButtox renders with disabled 1`] = `
<label
  className="ax-chedio-buttox__container ax-chedio-buttox__container--disabled ax-space--x2"
  onClick={undefined}
  title="Lorem ipsum dolor sit amet"
>
  <input
    className="ax-chedio-buttox ax-ipsum"
    disabled={true}
    type="checkbox"
  />
  <span
    className="ax-chedio-buttox__indicator ax-lorem"
  />
  <span
    className="ax-chedio-buttox__label"
  >
    Lorem ipsum
  </span>
</label>
`;

exports[`ChedioButtox renders with invalid 1`] = `
<label
  className="ax-chedio-buttox__container ax-chedio-buttox__container--invalid ax-space--x2"
  onClick={undefined}
  title="Lorem ipsum dolor sit amet"
>
  <input
    className="ax-chedio-buttox ax-ipsum"
    disabled={undefined}
    type="checkbox"
  />
  <span
    className="ax-chedio-buttox__indicator ax-lorem"
  />
  <span
    className="ax-chedio-buttox__label"
  >
    Lorem ipsum
  </span>
</label>
`;
